README.md
# Polaroid Carousel

A unique Flutter package for creating a polaroid-style carousel with item animations that slide from front to back or back to front. Add a touch of nostalgia to your Flutter app by using this package to display your content in a fun and interactive way.

![Polaroid Carousel Demo](https://firebasestorage.googleapis.com/v0/b/github-895c7.appspot.com/o/package%2Fpolaroid_carousel%2FFlutter%20Demo%20-%20Google%20Chrome%202023-10-13%2022-19-50.gif?alt=media&token=42c5c281-6a3f-44cf-a1cb-ad75d06939f3&_gl=1*10qsoge*_ga*MjA3OTcyNTA1Ni4xNjkzNzQxMzYy*_ga_CW55HF8NVT*MTY5NzIxNTkxNi4yOC4xLjE2OTcyMTU5ODYuNjAuMC4w)

## Features

- Front-to-back and back-to-front item sliding animations using order property.
- Customizable carousel item rotation (x,y,z) and translation direction (left,right,top,bottom).
- Smooth animation curves for a delightful user experience.

## Installation

To use this package, follow these steps:

1. Add the latest version of the package to your `pubspec.yaml` file and run `dart pub get`:

```
   dependencies:
     polaroid_carousel: ^1.0.0
 ```   

2. Import the package in your Dart code and use it in your Flutter app:

```
   import 'package:polaroid_carousel/polaroid_carousel.dart';
```

## Usage

Here's an example of how to use the PolaroidCarousel widget in your Flutter application:

- This widget accepts a list of widgets and displays them in animation. If you desire a specific rotation for each item, wrap each child with a transform and specify the desired rotation angle using rotateZ() like in [Example](example.dart).

```dart
class Example extends StatelessWidget {
  const Example({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    List<Widget> list = [
      Container(
        height: 200,
        width: 300,
        color: Colors.black,
      ),
      Container(
        height: 200,
        width: 300,
        color: Colors.blue,
      ),
      Container(
        height: 200,
        width: 300,
        color: Colors.brown,
      ),
      Container(
        height: 200,
        width: 300,
        color: Colors.red,
      ),
      Container(
        height: 200,
        width: 300,
        color: Colors.grey,
      ),
      Container(
        height: 200,
        width: 300,
        color: Colors.yellow,
      ),
    ];

    return Scaffold(
      body: Center(
          child: PolaroidCarousel(
          //these to property are necessary

          // value to which list item translate
          // advice : if the children are rotated like the example and you are not able to specify the translateFactor
          // then the max translateFactor should be √((childHeight^2, childWidth^2)) of child with maximum size
          //but try to adjust translateFactor according to your need
          translateFactor: 360,

          children: list,

          // these properties are optional

          // duration: const Duration(seconds: 1),
          // rotate: const Rotate(x: 0.001,y: 0.02,z: 0.001),
          // curve: Curves.easeInOut,
          // order: Order.frontToBack,
          // translate: Translate.top,
          ),
      ),
    );
  }
}
```

## Examples

Check out the [Example](example.dart) for different use cases and ideas on how to use the PolaroidCarousel widget in your app.

## Detail Explanation
You can find the full article on this package on [Medium](https://medium.com/@hashmack007/creating-dynamic-card-carousels-with-polaroid-carousel-in-flutter-702545b933dd).

## Contributing
If you find a bug or have a feature request, please open an issue. Contributions are welcome!

## License
This package is licensed under the MIT License. See the [LICENSE](LICENSE) file for details. 

CHANGELOG.md
## [1.0.0] - 2023-10-13

- Initial release of the "polaroid_carousel" package.
- Added a unique carousel widget that mimics a deck of polaroid cards.
- Supports sliding items from front to back and back to front.
- Provides options for customizing animation duration, rotation, order, and translations direction.


test/polaroid_carousel_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:polaroid_carousel/polaroid_carousel.dart';


void main() {
  test('polaroid_carousel', () {
    const PolaroidCarousel(translateFactor: 0,
    children: [],);
  });
}


lib/src/polaroid_carousel.dart
import 'package:flutter/material.dart';

class Rotate {
  final double x;
  final double y;
  final double z;

  /// Value should be less like 0.001 for translationFactor 360 because rotate value depends on [translationFactor]
  const Rotate({
    double x = 0.0,
    double y = 0.0,
    double z = 0.0,
  })  : assert(x >= 0.0 && x <= 1.0, 'x must be between 0.0 and 1.0'),
        assert(y >= 0.0 && y <= 1.0, 'y must be between 0.0 and 1.0'),
        assert(z >= 0.0 && z <= 1.0, 'z must be between 0.0 and 1.0'),
        x = (x >= 0.0 && x <= 1.0) ? x : (x < 0.0 ? 0.0 : 1.0),
        y = (y >= 0.0 && y <= 1.0) ? y : (y < 0.0 ? 0.0 : 1.0),
        z = (z >= 0.0 && z <= 1.0) ? z : (z < 0.0 ? 0.0 : 1.0);
}

/// Decide the animation direction - [frontToBack] or [backToFront]
enum Order {
  frontToBack,
  backToFront,
}

/// Translation direction [left] or [right] or [top] or [bottom]
enum Translate {
  left,
  right,
  top,
  bottom,
}

class PolaroidCarousel extends StatefulWidget {
  final Order order;
  final List<Widget> children;
  final double translateFactor;
  final Translate translate;
  final Curve curve;
  final Duration duration;
  final Rotate rotate;

  /// This unique carousel seamlessly shuffles items from the front and back, creating an engaging user interface.
  const PolaroidCarousel(
      {Key? key,
        required this.children,
        this.order = Order.frontToBack,
        this.translate = Translate.right,
        required this.translateFactor,
        this.curve = Curves.easeInSine,
        this.duration = const Duration(seconds: 2),
        this.rotate = const Rotate()})
      : super(key: key);

  @override
  State<PolaroidCarousel> createState() => _PolaroidCarouselState();
}

class _PolaroidCarouselState extends State<PolaroidCarousel> with TickerProviderStateMixin {
  bool order = false;
  late Widget topWidget;
  late List<Widget> children;
  late AnimationController controller;
  late Animation<double> animation;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    children = widget.children;
    order = widget.order == Order.backToFront;
    createAnimation();

    order
        ? {
      topWidget = children.removeAt(0),
      children.insert(0, animatedWidget(topWidget)),
    }
        : {
      topWidget = children.removeLast(),
      children.add(animatedWidget(topWidget)),
    };

    controller.forward();
    controller.addStatusListener(listen);
  }

  void createAnimation() {
    controller = AnimationController(vsync: this, duration: widget.duration);
    final curvedAnimation = CurvedAnimation(
      parent: controller,
      curve: widget.curve,
    );
    animation =
        Tween(begin: 0.0, end: widget.translateFactor).animate(curvedAnimation);
  }

  void renewController() {
    controller.dispose();

    createAnimation();

    order
        ? {
      children.removeLast(),
      children.add(topWidget),
      topWidget = children.removeAt(0),
      children.insert(0, animatedWidget(topWidget)),
    }
        : {
      children.removeAt(0),
      children.insert(0, topWidget),
      topWidget = children.removeLast(),
      children.add(animatedWidget(topWidget)),
    };

    setState(() {});

    controller.addStatusListener(listen);
    controller.forward();
  }

  void listen(AnimationStatus status) {
    if (status == AnimationStatus.completed) {
      // Animation has completed, reverse it
      order
          ? children.add(children.removeAt(0))
          : children.insert(0, children.removeLast());
      controller.reverse();
      setState(() {});
    } else if (status == AnimationStatus.dismissed) {
      // Animation has reversed back to the beginning, restart it
      renewController();
    }
  }

  Widget animatedWidget(Widget listItem) {
    return AnimatedBuilder(
      builder: (context, child) {
        return Transform(
          transform: Matrix4.identity()
            ..translate(
                widget.translate == Translate.left
                    ? -animation.value
                    : widget.translate == Translate.right
                    ? animation.value
                    : 0,
                widget.translate == Translate.top
                    ? -animation.value
                    : widget.translate == Translate.bottom
                    ? animation.value
                    : 0)
            ..rotateX(animation.value * widget.rotate.x)
            ..rotateY(animation.value * widget.rotate.x)
            ..rotateZ(animation.value * widget.rotate.x),
          child: listItem,
        );
      },
      animation: animation,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: children,
    );
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    controller.dispose();
  }
}


example.dart
import 'package:flutter/material.dart';
import 'dart:math' show pi;
import 'package:polaroid_carousel/polaroid_carousel.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const Example(),
    );
  }
}

class Example extends StatelessWidget {
  const Example({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    List<Widget> list = [
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(pi / 8),
        height: 200,
        width: 300,
        color: Colors.black,
      ),
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(pi / 40),
        height: 200,
        width: 300,
        color: Colors.blue,
      ),
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(pi / 15),
        height: 200,
        width: 300,
        color: Colors.brown,
      ),
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(-pi / 15),
        height: 200,
        width: 300,
        color: Colors.red,
      ),
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(-pi / 8),
        height: 200,
        width: 300,
        color: Colors.grey,
      ),
      Container(
        transformAlignment: Alignment.center,
        transform: Matrix4.identity()..rotateZ(-pi / 40),
        height: 200,
        width: 300,
        color: Colors.yellow,
      ),
    ];

    return Scaffold(
      body: Center(
          child: PolaroidCarousel(
        //these to property are necessary

        // value to which list item translate
        // advice : if the children are rotated like the example and you are not able to specify the translateFactor
        // then the max translateFactor should be √((childHeight^2, childWidth^2)) of child with maximum size
        //but try to adjust translateFactor according to your need
        translateFactor: 360,

        children: list,

        // these properties are optional

        // duration: const Duration(seconds: 1),
        // rotate: const Rotate(x: 0.001,y: 0.02,z: 0.001),
        // curve: Curves.easeInOut,
        // order: Order.frontToBack,
        // translate: Translate.top,
      )),
    );
  }
}
